### PR TITLE
ATDD : ATDgen backend to generate `dlang` code

### PR DESCRIPTION
**Pull request by elrandar**
# ATDD
This PR adds the `atdd` backend, generating `dlang` serializers and deserializers from ATD definitions. Supported features are:
- optionals
- default values
- sumtypes
- wrapping and unwrapping with annotations
- specifying dlang imports with top level annotations eg: `<dlang import="std.stdint : uint32_t">`

Also adds unit tests for testing and checking compilation of generated code. To execute those the `ldc` compiler is needed, more information in `readme.md`.

### PR checklist

- [x] New code has tests to catch future regressions
- [ ] Documentation is up-to-date
- [x] `CHANGES.md` is up-to-date

### CODE REVIEW

File: atdd/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+Development notes
+--
+
+Build or rebuild with `make`. Test with `make test`. This requires
+ldc2.
+
+Running the tests is done from the `atdd/` main folder with `make
+test`.
+
+We have two kinds of tests for atdpy:
```
Comment (Khady) [inline atdd/README.md:43]:
```suggestion
We have two kinds of tests for atdd:
```

File: atdd/src/bin/Atdd_main.ml [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  "Type-safe JSON serializers for Python"
+
+(*
+   The structure of the help page.
+*)
+let man = [
+  (* 'NAME' and 'SYNOPSIS' sections are inserted here by cmdliner. *)
+
+  `S Manpage.s_description;  (* standard 'DESCRIPTION' section *)
+  `P "atdd turns a file containing type definitions into Python classes \
```
Comment (Khady) [inline atdd/src/bin/Atdd_main.ml:58]:
```suggestion
  `P "atdd turns a file containing type definitions into D classes \
```

File: atdd/src/bin/Atdd_main.ml [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+      that read, write, and validate JSON data. The generated code \
+      can be type-checked statically with mypy to ensure user code agrees \
+      with the ATD interface.";
+
+  (* 'ARGUMENTS' and 'OPTIONS' sections are inserted here by cmdliner. *)
+
+  `S Manpage.s_examples; (* standard 'EXAMPLES' section *)
+  `P "The following is a sample ATD file. 'sample.atd' becomes 'sample.py' \
```
Comment (Khady) [inline atdd/src/bin/Atdd_main.ml:66]:
```suggestion
  `P "The following is a sample ATD file. 'sample.atd' becomes 'sample.d' \
```

File: atdd/src/bin/Atdd_main.ml [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      with the command 'atdd sample.atd'.";
+  `Pre "\
+(* Sample ATD file sample.atd *)
+
+type foo = {
+  name: string;                 (* required field *)
+  ?description: string option;  (* optional field *)
+  ~tags: string list;           (* optional with implicit default *)
+  ~price <python default=\"0.99\">: float;  (* explicit default *)
```
Comment (Khady) [inline atdd/src/bin/Atdd_main.ml:75]:
```suggestion
  ~price <dlang default=\"0.99\">: float;  (* explicit default *)
```

File: atdd/test/.gitignore [hunk 5]
```diff
@@ -0,0 +1,5 @@
+__pycache__
```
Comment (Khady) [inline atdd/test/.gitignore:1]: can be dropped?

File: atdd/test/.gitignore [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+
+!dlang-expected/
+dlang-tests/*.d
+dlang-tests/*.py
```
Comment (Khady) [inline atdd/test/.gitignore:5]: same?

**Review by Khady — COMMENTED** [review]
generally looks good to me and I think that we can merge even if the full doc isn't present.
`ldc` may be added as an external dependencies for test in the opam file?

**Comment by Khady** [comment]
git history can probably be squashed?

File: CHANGES.md [hunk 7]
```diff
@@ -1,3 +1,7 @@
+master
+-------------------
+* atdd: Add `dlang` backend to generate D code from ATD definitions
```
Comment (Khady) [inline CHANGES.md:3]:
```suggestion
* atdd: Add `dlang` backend to generate D code from ATD definitions (#349)
```

File: atdd.opam [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+homepage: "https://github.com/ahrefs/atd"
+bug-reports: "https://github.com/ahrefs/atd/issues"
+depends: [
+  "dune" {>= "2.8"}
+  "ocaml" {>= "4.08"}
+  "atd" {>= "2.11.0"}
+  "cmdliner" {>= "1.1.0"}
+  "re"
+  "alcotest" {with-test}
+  "odoc" {with-doc}
```
Comment (Khady) [inline atdd.opam:70]: @elrandar I noticed that you updated this file by hand to remove `alcotest` and to add `ldc2`. This file is generated, it shouldn't be updated by hand. Check `dune-project` instead.

**Review by Khady — APPROVED** [review]
LGTM, thanks

**Review by mjambon — APPROVED** [review]
I haven't looked closely but it looks good in spirit. Great job.

**Comment by Khady** [comment]
I forgot to switch to `squash` when I merged the PR, sorry about that. If we want to cleanup the history the master branch could be forced pushed

**Comment by mjambon** [comment]
A heads-up: I'm planning to release atd (all opam packages) tomorrow or early next week. This will include atdd.

**Pull request merged**